MapboxDirections/MBVisualInstructionDetailComponent.swift

```diff
@@ -0,0 +1,93 @@
+import Foundation
+
+
+/**
+ :nodoc:
+ Contains necessary features to build a more rich visual experience for a `RouteStep`.
+ */
+@objc(MBVisualInstructionDetailComponent)
+public class VisualInstructionDetailComponent: NSObject, NSSecureCoding {
+    
+    /**
+     :nodoc:
+     Single part of a visual instruction.
+     */
+    public let text: String?
+    
+    
+    /**
+     :nodoc:
+     @1x PNG to inline in visual instruction.
+     */
+    public var imageURL1x: URL? = nil
+    
+    
+    /**
+     :nodoc:
+     @2x PNG to inline in visual instruction.
+     */
+    public var imageURL2x: URL? = nil
```

Comment:
@frederoni @1ec5 @JThramer what do you think of this api design?

Reply:
Well if this was swift-only, I think this should be encapsulated in a named tuple or a struct, but OBJ-C continues to bind our hands in this regard (until we start [playing with](http://nshipster.com/new-years-2016/) `__ObjCConvertable`, anyway...)

How does the rest of the class feel about using objects to encapsulate closely-related properties? It will be way less of a breaking change for an API user to go from Object -> Struct down the road then from 4 or 5 top-level properties into a struct.

Reply:
Combine the various `imageURL*x` properties into a single dictionary mapping UITraitCollections (created using `UITraitCollection(displayScale:)`) to image URLs. This will make it easier for the client to build a UIImageAsset from the various URLs if desired, and it gives us the flexibility to vend different images based on orientation or other traits in the future. A `[UITraitCollection: URL]` also bridges to Objective-C as is.

Reply:
Noting, this will require importing UIKit.

Reply:
Conditionally import UIKit for iOS and tvOS.

For macOS, it’s possible to create an NSImage that has representations at multiple scales, and it’s also possible to create an NSImage lazily loads a URL using `NSImage(representingUrl:)`, but I don’t think it’s possible to combine the two features. So `[NSNumber: URL]` appears to be our only option on that platform.